File: app/src/main/java/at/shockbytes/corey/fragment/ScheduleFragment.java
package at.shockbytes.corey.fragment;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import at.shockbytes.corey.R;
import at.shockbytes.corey.adapter.DaysScheduleAdapter;
import at.shockbytes.corey.adapter.ScheduleAdapter;
import at.shockbytes.corey.adapter.helper.CoreyItemTouchHelper;
import at.shockbytes.corey.core.CoreyApp;
import at.shockbytes.corey.fragment.dialogs.InsertScheduleDialogFragment;
import at.shockbytes.corey.storage.live.LiveScheduleUpdateListener;
import at.shockbytes.corey.common.core.util.ResourceManager;
import at.shockbytes.corey.util.schedule.ScheduleItem;
import at.shockbytes.corey.util.schedule.ScheduleManager;
import at.shockbytes.corey.common.core.util.view.EqualSpaceItemDecoration;
import butterknife.Bind;
import butterknife.ButterKnife;
import rx.functions.Action1;

/**
 * @author Martin Macheiner
 *         Date: 26.10.2015.
 */
public class ScheduleFragment extends Fragment implements LiveScheduleUpdateListener,
        ScheduleAdapter.OnItemMoveListener, ScheduleAdapter.OnItemClickListener,
        InsertScheduleDialogFragment.OnScheduleItemSelectedListener {


    public static ScheduleFragment newInstance() {
        ScheduleFragment fragment = new ScheduleFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private ItemTouchHelper touchHelper;

    private ScheduleAdapter adapter;

    @Inject
    protected ScheduleManager scheduleManager;

    @Bind(R.id.fragment_schedule_rv)
    protected RecyclerView recyclerView;

    @Bind(R.id.fragment_schedule_rv_days)
    protected RecyclerView recyclerViewDays;

    public ScheduleFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((CoreyApp) getActivity().getApplication()).getAppComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_schedule, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onStart() {
        super.onStart();
        initializeViews();
    }

    @Override
    public void onStop() {
        super.onStop();
        scheduleManager.unregisterLiveForScheduleUpdates(this);
    }

    @Override
    public void onScheduleItemAdded(ScheduleItem item) {
        adapter.insertEntity(item);
    }

    @Override
    public void onScheduleItemDeleted(ScheduleItem item) {
        adapter.resetEntity(item);
    }

    @Override
    public void onScheduleItemChanged(ScheduleItem item) {
        adapter.updateEntity(item);
    }

    @Override
    public void onItemMove(ScheduleItem item, int from, int to) {

    }

    @Override
    public void onItemMoveFinished() {

        List<ScheduleItem> items = adapter.getScheduleData();
        for (ScheduleItem si : items) {
            scheduleManager.updateScheduleItem(si);
        }
    }

    @Override
    public void onItemDismissed(ScheduleItem item, int position) {

        if (!item.isEmpty()) {
            scheduleManager.deleteScheduleItem(item);
        }
        //adapter.addEntity(position, new ScheduleItem("", position));
        //recyclerView.invalidate();
    }

    @Override
    public void onItemClick(ScheduleItem item, View v, int position) {

        if (item.isEmpty()) {
            InsertScheduleDialogFragment fragment = InsertScheduleDialogFragment
                    .newInstance(position);
            fragment.setOnScheduleItemSelectedListener(this);
            fragment.show(getFragmentManager(), fragment.getTag());
        }
    }

    @Override
    public void onScheduleItemSelected(String item, int day) {
        scheduleManager.insertScheduleItem(new ScheduleItem(item, day));
    }

    private void initializeViews() {

        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), getColsForOrientation()));
        recyclerViewDays.setLayoutManager(new GridLayoutManager(getContext(), getColsForOrientation()));
        recyclerViewDays.setAdapter(new DaysScheduleAdapter(getContext(),
                Arrays.asList(getResources().getStringArray(R.array.days))));
        recyclerViewDays.addItemDecoration(new EqualSpaceItemDecoration(
                ResourceManager.convertDpInPixel(4, getContext())));

        adapter = new ScheduleAdapter(getContext(), new ArrayList<ScheduleItem>());
        recyclerView.setAdapter(adapter);
        recyclerView.setNestedScrollingEnabled(false);
        recyclerView.addItemDecoration(new EqualSpaceItemDecoration(
                ResourceManager.convertDpInPixel(4, getContext())));
        ItemTouchHelper.Callback callback = new CoreyItemTouchHelper(adapter, true, false);
        adapter.setOnItemMoveListener(this);
        adapter.setOnItemClickListener(this);

        touchHelper = new ItemTouchHelper(callback);
        touchHelper.attachToRecyclerView(recyclerView);

        scheduleManager.getSchedule().subscribe(new Action1<List<ScheduleItem>>() {

            @Override
            public void call(List<ScheduleItem> scheduleItems) {
                adapter.setData(scheduleItems);
            }
        }, new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                Toast.makeText(getContext(), throwable.toString(), Toast.LENGTH_LONG).show();
                throwable.printStackTrace();
            }
        });

        scheduleManager.registerLiveForScheduleUpdates(this);
    }

    private int getColsForOrientation() {
        return getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT
                ? 1
                : 7;
    }

}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/pager/ScheduleFragment.kt
package at.shockbytes.corey.ui.fragment.pager

import android.content.res.Configuration
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.HapticFeedbackConstants
import android.widget.Toast
import at.shockbytes.core.ui.fragment.BaseFragment
import at.shockbytes.corey.R
import at.shockbytes.corey.ui.adapter.DaysScheduleAdapter
import at.shockbytes.corey.ui.adapter.ScheduleAdapter
import at.shockbytes.corey.common.addTo
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.data.schedule.ScheduleItem
import at.shockbytes.corey.data.schedule.ScheduleRepository
import at.shockbytes.corey.ui.fragment.dialog.InsertScheduleDialogFragment
import at.shockbytes.util.AppUtils
import at.shockbytes.util.adapter.BaseAdapter
import at.shockbytes.util.adapter.BaseItemTouchHelper
import at.shockbytes.util.view.EqualSpaceItemDecoration
import kotlinx.android.synthetic.main.fragment_schedule.*
import kotterknife.bindView
import timber.log.Timber
import javax.inject.Inject

/**
 * Author:  Martin Macheiner
 * Date:    26.10.2015
 */
class ScheduleFragment : BaseFragment<AppComponent>(), BaseAdapter.OnItemMoveListener<ScheduleItem> {

    override val snackBarBackgroundColorRes: Int = R.color.sb_background
    override val snackBarForegroundColorRes: Int = R.color.sb_foreground

    @Inject
    lateinit var scheduleRepository: ScheduleRepository

    private lateinit var touchHelper: ItemTouchHelper
    private val adapter: ScheduleAdapter by lazy {
        ScheduleAdapter(
                context!!,
                { item, _, position -> onScheduleItemClicked(item, position) },
                { item, position -> onItemDismissed(item, position) }
        )
    }

    private val recyclerView: RecyclerView by bindView(R.id.fragment_schedule_rv)
    private val recyclerViewDays: RecyclerView by bindView(R.id.fragment_schedule_rv_days)

    private val recyclerViewLayoutManager: RecyclerView.LayoutManager
        get() = if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        } else {
            GridLayoutManager(context, ScheduleAdapter.MAX_SCHEDULES)
        }

    override val layoutId = R.layout.fragment_schedule

    override fun onItemMove(t: ScheduleItem, from: Int, to: Int) = Unit

    override fun onItemMoveFinished() {
        adapter.reorderAfterMove()
                .forEach { item ->
                    scheduleRepository.updateScheduleItem(item)
                }
    }

    override fun onItemDismissed(t: ScheduleItem, position: Int) {
        if (!t.isEmpty) {
            scheduleRepository.deleteScheduleItem(t)
        }
    }

    override fun bindViewModel() {

        scheduleRepository.schedule
                .subscribe({ scheduleItems ->
                    adapter.updateData(scheduleItems)
                }, { throwable ->
                    throwable.printStackTrace()
                    Toast.makeText(context, throwable.toString(), Toast.LENGTH_LONG).show()
                })
                .addTo(compositeDisposable)
    }

    override fun unbindViewModel() = Unit

    override fun setupViews() {

        context?.let { ctx ->

            recyclerViewDays.layoutManager = recyclerViewLayoutManager
            recyclerViewDays.adapter = DaysScheduleAdapter(ctx, resources.getStringArray(R.array.days).toList())
            recyclerViewDays.addItemDecoration(EqualSpaceItemDecoration(AppUtils.convertDpInPixel(4, ctx)))

            recyclerView.layoutManager = recyclerViewLayoutManager
            recyclerView.isNestedScrollingEnabled = false
            recyclerView.addItemDecoration(EqualSpaceItemDecoration(AppUtils.convertDpInPixel(4, ctx)))
            val callback = BaseItemTouchHelper(adapter, false, BaseItemTouchHelper.DragAccess.ALL)
            adapter.onItemMoveListener = this

            touchHelper = ItemTouchHelper(callback)
            touchHelper.attachToRecyclerView(recyclerView)
            recyclerView.adapter = adapter
        }

        fragment_schedule_card_clear_all?.setOnClickListener { view ->
            view.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
            showRemoveApprovalDialog()
        }
    }

    private fun showRemoveApprovalDialog() {
        context?.let { ctx ->
            AlertDialog.Builder(ctx)
                    .setTitle(R.string.delete_schedule)
                    .setMessage(R.string.delete_schedule_message)
                    .setIcon(R.drawable.ic_cancel_red)
                    .setNegativeButton(R.string.cancel) { _, _ -> Unit }
                    .setPositiveButton(R.string.delete) { _, _ ->
                        scheduleRepository.deleteAll()
                                .subscribe({
                                    Timber.d("All items deleted")
                                }, { throwable ->
                                    showToast("Could not delete all items: ${throwable.message}")
                                })
                                .addTo(compositeDisposable)
                    }
                    .create()
                    .show()
        }
    }

    override fun injectToGraph(appComponent: AppComponent?) {
        appComponent?.inject(this)
    }

    private fun onScheduleItemClicked(item: ScheduleItem, position: Int) {
        if (item.isEmpty) {
            InsertScheduleDialogFragment.newInstance()
                    .setOnScheduleItemSelectedListener { i ->
                        scheduleRepository.insertScheduleItem(ScheduleItem(i.item.title, position, locationType = i.item.locationType))
                    }
                    .show(fragmentManager, "dialogfragment-insert-schedule")
        }
    }

    companion object {

        fun newInstance(): ScheduleFragment {
            val fragment = ScheduleFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
